pkg/plugin/processor/builtin/unwrap_opencdc.go

```diff
@@ -0,0 +1,290 @@
+
+	switch opType := op.(type) {
+	case opencdc.Operation:
+		operation = opType
+	case string:
+		if err := operation.UnmarshalText([]byte(opType)); err != nil {
+			return operation, cerrors.Errorf("invalid operation %q", opType)
+		}
+	default:
+		return operation, cerrors.Errorf("expected a opencdc.Operation or a string, got %T", opType)
+	}
+	return operation, nil
+}
+
+// unmarshalMetadata extracts metadata from a structuredData record.
+func (u *unwrapOpenCDC) unmarshalMetadata(structData opencdc.StructuredData) (opencdc.Metadata, error) {
+	var metadata opencdc.Metadata
+	meta, ok := structData["metadata"]
+	if !ok {
+		return metadata, cerrors.Errorf("record payload after doesn't contain metadata")
+	}
+
+	switch m := meta.(type) {
+	case opencdc.Metadata:
+		metadata = m
+	case map[string]interface{}:
+		metadata = make(opencdc.Metadata, len(m))
+		for k, v := range m {
+			metadata[k] = fmt.Sprint(v)
+		}
+	default:
+		return metadata, cerrors.Errorf("expected a opencdc.Metadata or a map[string]interface{}, got %T", m)
+	}
+	return metadata, nil
+}
+
+// unmarshalKey extracts key from a structuredData record.
+func (u *unwrapOpenCDC) unmarshalKey(structData opencdc.StructuredData) (opencdc.Data, error) {
+	var key opencdc.Data
+	ky, ok := structData["key"]
+	if !ok {
+		return key, cerrors.Errorf("record payload after doesn't contain key")
+	}
+	switch k := ky.(type) {
+	case map[string]interface{}:
+		convertedData := make(opencdc.StructuredData, len(k))
+		for kk, v := range k {
+			convertedData[kk] = v
+		}
+		key = convertedData
+	case string:
+		decoded := make([]byte, base64.StdEncoding.DecodedLen(len(k)))
+		n, err := base64.StdEncoding.Decode(decoded, []byte(k))
+		if err != nil {
+			return key, cerrors.Errorf("couldn't decode key: %w", err)
+		}
+		key = opencdc.RawData(decoded[:n])
+	default:
+		return key, cerrors.Errorf("expected a opencdc.Data or a string, got %T", k)
+	}
+	return key, nil
+}
+
+func (u *unwrapOpenCDC) convertPayloadData(payload map[string]interface{}, key string) (opencdc.Data, error) {
+	payloadData, ok := payload[key]
+	if !ok {
+		return nil, nil
+	}
+
+	switch data := payloadData.(type) {
+	case map[string]interface{}:
+		convertedData := make(opencdc.StructuredData, len(data))
+		for k, v := range data {
+			convertedData[k] = v
+		}
+		return convertedData, nil
+	case string:
+		decoded := make([]byte, base64.StdEncoding.DecodedLen(len(data)))
+		n, err := base64.StdEncoding.Decode(decoded, []byte(data))
+		if err != nil {
+			return nil, cerrors.Errorf("couldn't decode payload %s: %w", err, key)
+		}
+		return opencdc.RawData(decoded[:n]), nil
+	default:
+		return nil, nil
+	}
+}
+
+// unmarshalPayload extracts payload from a structuredData record.
+func (u *unwrapOpenCDC) unmarshalPayload(structData opencdc.StructuredData) (opencdc.Change, error) {
+	var payload opencdc.Change
+	pl, ok := structData["payload"]
+	if !ok {
+		return payload, cerrors.Errorf("record payload doesn't contain payload")
```

Comment:
this error msg might make sense for us, but the user will definitely be confused 👀

Reply:
True!